docs/RFCS/20171118_role_based_access_control.md

```diff
@@ -0,0 +1,757 @@
+* lookup all roles: `SHOW GRANTS ON ROLE *`
+* show all memberships for a user: `SHOW GRANT ON ROLE * FOR marc`
+* show user membership in a role: `SHOW GRANT ON ROLE myrole FOR marc`
+
+### Administrators
+
+Before this RFC, there is a single administrative user in a cockroach cluster: the user `root`.
+
+We propose to add a default role: `admin` with a single user: `root`.
+`root` will have admin privileges on the `admin` role, allowing addition/removal of other users.
+
+The `admin` role cannot be deleted, and cannot be empty. The special `root` user cannot be dropped or removed
+from the `admin` role.
+
+All operations that previously checked for the `root` user will now check for membership in the `admin` role.
+
+Clusters created before the existence of roles will have a new `admin` role created automatically and privileges
+for it added where `root` currently has privileges. `root` will keep its individual privileges to allow downgrades
+to older versions.
+
+### Non-enterprise users
+
+Manipulation of roles is an enterprise feature.
+
+Without a valid enterprise license, the following are not allowed:
+* creating a role (`CREATE ROLE ...`)
+* dropping a role (`DROP ROLE ...`)
+* adding a member of a role (`GRANT myrole TO ...`)
+* removing a member from a role (`REVOKE myrole TO ...`)
+
+The following will function without an enterprise license:
+* listing roles
+* listing role memberships
+* granting privileges to a role
+* revoking privileges from a role
+
+This restriction leaves non-enterprise users with a fully-functioning product, but without the
+ease of management provided by roles.
+
+## Contributor impact
+
+Any modifications to the SQL code must keep in mind:
+* privileges can be applied to roles and users
+* roles and users share the same namespace
+
+# Reference-level explanation
+
+A number of concepts are detailed in the [Guide-level explanation](#guide-level-explanation) and not
+repeated in this section.
+
+## Detailed design
+
+### Roles system table
+
+We propose the modification of the users table to store roles, and the addition of a new table to store
+role memberships.
+
+#### Users table
+
+Since roles and users share the same namespace, it is simplest to alter the `system.users` table to
+include roles.
+
+The schema for the users table becomes:
+The `role` table stores the list of all existing roles.
+```
+CREATE TABLE system.users (
+  username         STRING PRIMARY KEY,
+  "hashedPassword" BYTES,
+  isRole           BOOL,
+  (INDEX isRole)
+);`
+```
+
+Depending on the complexity of migration, we may be able to rename the `username` field and even the table.
+
+`isRole` denotes whether the corresponding name is a user or a role. Roles cannot log in or have passwords.
+
+#### Role members table
+
+The `role_members` stores the list of all role memberships. This only stores direct relationships:
+
+```
+CREATE TABLE system.role_members (
+  role          STRING,
+  member        STRING,
```

Comment:
Using Foreign Keys here will make it simpler to manage. Do we want to use Foreign Keys? We would need to have two columns user_member and role_member (a self referential key)

Reply:
Or should we be using the OID type to reference these?

Reply:
Self-referential keys are currently a problem and don't really add anything. We have plenty of checks and cleanup to do when dropping a user, removing its `role_members` entries is pretty straightforward.

AFAICT, OIS is just an alias used in the virtual tables. The virtual tables section uses those, but I don't think we need them internally.